File: Pepcoding Classes/Foundation/Arrays and 2D array/l003array.java
import java.util.Scanner;
import java.util.Arrays;

public class l003array{
    public static Scanner scn = new Scanner(System.in);

    // public static void arrayInput(){
    //     int[] arr = new int[10];
    //     for(int i=0; i < 20; i++){
    //         arr[i] = scn.nextInt();
    //     } 
    // }



    // Basic Question ------------------------------
    // Find the maximum value
    public static int maximumElement(int[] arr){
        int maxEle = (int)-1e8;
        for(int i= 0;i< arr.length ; i++){
            if(arr[i] > maxEle){
                maxEle = arr[i];
            }
        }
        return maxEle;
    }

    // Find the minmum value
    public static int minmumElement(int[] arr){
        int minEle = (int)1e8;
        for(int i= 0;i< arr.length ; i++){
            if(arr[i] < minEle){
                minEle = arr[i];
            }
        }
        return minEle;
    }

    // Find the entered element
    public static int findElement(int[] arr,int data){
        int idx = -1;
        for(int i= 0;i< arr.length ; i++){
            if(arr[i] == data){
                idx = i;
                break;
            }
        }
        return idx;
    }

    // Find the first index of entered element
    public static int firstIndex(int[] arr,int data){
        int idx = -1;
        for(int i= 0;i< arr.length ; i++){
            if(arr[i] == data){
                idx = i;
                break;
            }
        }
        return idx;
    }

    // Find the last index of entered element
    public static int lastIndex(int[] arr,int data){
        int idx = -1;
        for(int i= arr.length - 1;i >= 0; i--){
            if(arr[i] == data){
                idx = i;
                break;
            }
        }
        return idx;
    }

    // -----------------------------------------------------------------------------

    public static void sumOfTwoArrays(int[] arr1, int[] arr2){
        int p = arr1.length;
        int q = arr2.length;
        int r = Math.max(p,q) + 1;

        int[] arr = new int[r];

        int carry = 0 ;
        int i = p -1;
        int j = q -1;
        int k = r -1;


        while(k >= 0){
            int sum = carry;
            if(i >= 0){
                sum += arr1[i--];
            }    
            if(j >= 0){
                sum += arr2[j--];
            }    

            ans[k--] = sum % 10;
            carry = sum / 10;
        }

        for(int l = 0; l < ans.length ; l++){
            if( l == 0 && ans[l] == 0){
                continue;
            } 
            System.out.println(ans[l]);
        }
    }

    public static void subOfTwoArrays(int[] arr1, int[] arr2){
        int p = arr1.length;
        int q = arr2.length;
        int r = p;

        int[] ans = new int[r];

        int borrow = 0 ;
        int i = p -1;
        int j = q -1;
        int k = r -1;


        while(k >= 0){
            int num = borrow;
            if(i >= 0){
                num += arr1[i--];
            }    
            if(j >= 0){
                num -= arr2[j--]; // Subtract
            }    
            
            if(num < 0){
                num += 10;
                borrow = -1;
            } else {
                borrow = 0;
            }
            ans[k--] = num;
        }
        boolean flag = false;
        for(int l = 0; l < ans.length ; l++){
            if( !flag && ans[l] == 0){
                continue;
            } 
            flag = true;
            System.out.println(ans[l]);
        }
    }

    public static void inverse(int[] arr){
        int n = arr.length;
        int[] ans = new int[n];

        for(int i = 0; i < ans.length; i++){
            ans[arr[i]] = i ;
        }

        for(int ele : ans){
            System.out.println(ele + " ");
        }
    }

    // -----------------------------------------------------------------------------------------
    //  Binary Search 
    public static int binarySearch(int[] arr, int key ){
        int start = 0;
        int end = arr.length - 1;
        while(start <= end){
            int mid = (start + end) / 2;
            if(key == arr[mid]){
                return mid;
            } else if(key < arr[mid]){
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return -1;
    }

    
    // -----------------------------------------------------------------------------------------
    public static void saddlepoint(int[][] arr){
        int max = -1e9;
        int min = 1e9;
        for(int i = 0; i < arr.length; i++){
            for(int j = 0; j < arr[0].length ; j++){
                if(arr[i][j] > max) max = arr[i][j];
                if(arr[j][i] < min) min = arr[j][i];    
            }
            if(min == max) System.out.print(max);
        }
    }


    // Diagonal Traversal 
    public static void diagonalTraverse(int[][] arr){
        for(int gap = 0 ; gap < arr[0].length ; gap++){
            for(int i = 0, j = gap; i < arr.length && j < arr[0].length ; i++,j++)
                System.out.println(arr[i][j]);
        }
    }

    public static void main(String[] args) throws Exception {
        // write your code here
        int n = scn.nextInt();
        int[][] arr = new int[n][n];
        for(int i= 0 ;i < arr.length; i++){
            for(int j = 0 ; j < arr[0].length ; j++){
                arr[i][j] = scn.nextInt();
            }
        }
        diagonalTraverse(arr);
    }


    public static void saddlepoint(int[][] arr){
        int n = arr.length;
        int m = arr[0].length;
        for(int r = 0; r < n ; r++){
            int c = 0;
            int minEle = (int)1e8;

            for(int j = 0; j < m ; j++){
                if(arr[r][j] < minEle){
                    minEle = arr[r][j];
                    c = j;
                }
            }

            boolean isSaddlePoint = true;

            for(int i = 0; i < n ; i++){
                if(arr[i][c] > minEle){
                    isSaddlePoint = false;
                    break;
                }
            }
            if(isSaddlePoint){
                System.out.print(minEle);
            }
        }
        System.out.print("Invalid input");
    }


    public static void searhcSortedMatrix(int[][] arr){
        int j = 0 ;
        int i = arr.length - 1;

        while(i >= 0 && j < arr.length )

            if(arr[i][j]) == data){
                System.out.print(i + "\n" + j);
                return;
            }
            else if(arr[i][j] < data){
                j++;
            }
            else{
                i--;
            }
    }

    
    public static void rotate90(int[][] arr){
        
        // Transpose
        int n = arr.length;
        int m = arr[0].length;
        for(int i = 0; i < n; i++){
            for(int j = i; j < m; j++){
                int temp = arr[i][j];
                arr[i][j] = arr[j][i] ; 
                arr[j][i] = temp ; 
            }
        }
        
        int c1 = 0;
        int c2 = n -1;
        while(c1 < c2){
            
            for(int i = 0 ; i < n; i++){
                int temp = arr[i][c1];
                arr[i][c1] = arr[i][c2] ; 
                arr[i][c2] = temp ; 
            }
            c1++;
            c2--;
        } 
        
    }
    
    public static void spiralDisplay(int[][] arr){

        int n = arr.length, m = arr[0].length, count = n*m;
        int rmin = 0, rmax = n -1;
        int cmin = 0, cmax = m -1;

        while(count != 0){
            
            for(int c = rmin; c < rmax && count > 0; c++){
                System.out.println(arr[cmin],c);
                count--;
            }
            cmin++;

            for(int r = cmin; r < cmax && count > 0; r++){
                System.out.println(arr[r],rmax);
                count--;
            }
            rmax--;

            for(int c = rmax; c >= rmax && count > 0; c++){
                System.out.println(arr[cmax],c);
                count--;
            }
            cmax--;

            for(int r = cmax; r >= cmin && count > 0 ; r++){
                System.out.println(arr[r],rmax);
                count--;
            }
            rmin++;        
        }
    }

    // Rotate a shell
    public static Scanner scn = new Scanner(System.in);
    
    public static void reverse(int[] arr,int i ,int j){
        while(i < j){
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
            j--;
        }
    }
    public static void rotate(int[] arr, int r){
        int n = arr.length;
        // if(r < 0){
        //     r = r + arr.length;
        // }
        r = (r %n + n) % n;
        // reverse(arr,0,n-1);
        // reverse(arr,0,n-r-1);
        // reverse(arr,n-r,n-1);
        reverse(arr,0,n-1);
        reverse(arr,0,r-1);
        reverse(arr,r,n-1);
    }
    
    public static int[] fillOneDfromShell(int[][] arr,int s){
        int n = arr.length;
        
        int rmin = s - 1;
        int cmin = s - 1;
        int rmax = n - s;
        int cmax = n - s;
        
        // int sz = lw + bw + rw + tw - 4 ; 
        // int sz = 2 *lw + 2 * bw - 4 ; 
        // int sz = 2 * (rmax - rmin + 1) + 2 * (cmax - cmin + 1) - 4 ; 
        int sz = 2*rmax - 2*rmin + 2*cmax - 2*cmin ; 
        
        int[] oned = new int[sz];
        
        // lw
        int idx = 0;
        for(int i = rmin , j = cmin; i <= rmax ; i++){
            oned[idx] = arr[i][j];
            idx++;
        }
        
        // bw
        for(int i = rmax, j = cmin + 1; j <= cmax ; j++){
            oned[idx] = arr[i][j];
            idx++;
        }
        
        // rw 
        for(int i = rmax - 1, j = cmax ; i >= rmin ; i--){
            oned[idx] = arr[i][j];
            idx++;
        }
        
        // tw
        for(int i = rmin , j = cmax - 1 ; j >= rmin + 1 ; j--){
            oned[idx] = arr[i][j];
            idx++;
        }
        return oned;
    }
    public static void fillShellFromOneD(int[][] arr,int s,int[] oned){
        int n = arr.length;
        int rmin = s - 1;
        int cmin = s - 1;
        int rmax = n - s;
        int cmax = n - s;
        
        
        // lw
        int idx = 0;
        for(int i = rmin , j = cmin; i <= rmax ; i++){
            arr[i][j] = oned[idx];
            idx++;
        }
        
        // bw
        for(int i = rmax, j = cmin + 1; j <= cmax ; j++){
            arr[i][j] = oned[idx];
            idx++;
        }
        
        // rw 
        for(int i = rmax - 1, j = cmax ; i >= rmin ; i--){
            arr[i][j] = oned[idx];
            idx++;
        }
        
        // tw
        for(int i = rmin , j = cmax - 1 ; j >= rmin + 1 ; j--){
            arr[i][j] = oned[idx];
            idx++;
        }
    }
    
    public static void shellrotate(int[][] arr,int s,int r){
        int[] oneD = fillOneDfromShell(arr,s);
        rotate(oneD,r);
        fillShellFromOneD(arr,s,oneD);
    }
    // public static void main(String[] args) throws Exception {
    //     // write your code here
    //     int n = scn.nextInt();
    //     int m = scn.nextInt();
    //     int[][] arr = new int[n][m];
    //     for(int i = 0; i < arr.length; i++){
    //         for(int j = 0; j < arr[0].length; j++){
    //             arr[i][j] = scn.nextInt(); 
    //         }
    //     }
    //     int s = scn.nextInt();
    //     int r = scn.nextInt();
        
    //     shellrotate(arr,s,r);
    //     display(arr);
    // }

    public static void display(int[][] arr){
        for(int i = 0; i < arr.length; i++){
            for(int j = 0; j < arr[0].length; j++){
                System.out.print(arr[i][j] + " ");
            }
            System.out.println();
        }
    }

}



// ------------------------------------------------------------


    public static void main(String[] args){

        int n = scn.nextInt();
        int[] arr = new int[n];
        for(int i = 0; i < arr.length ;i++){
            arr[i] = scn.nextInt();
        } 
        
        int data = scn.nextInt();
        // System.out.print(lastIndex(arr,data));
        System.out.print(firstIndex(arr,data));
    }
}

File: Pepcoding Classes/Foundation/Arrays and 2D array/l003array.cpp
#include<iostream>
#include<vector>
using namespace std;

void test1(){
    int* arr = new int[10];
    for(int i=0; i < 20; i++){
        cout << arr[i] << " ";
    } 
}
void test2(){
    vector<int> arr(10);
    for(int i=0; i < 20; i++){
        cout << arr[i] << " ";
    } 
}

void arrayInput(){
    int n;
    cin >> n;

    int* arr = new int[n];
    for(int i=0; i < n; i++){
        cout << arr[i] << " ";
    }
}

// Vector Input 
void vectorInput(){
    int n;
    cin >> n;

    vector<int> arr(n);
    for(int i=0; i < arr.size(); i++){
        cin >> arr[i];
    }
}



// Basic Question ------------------------------
// Find the maximum value
int maximumElement(vector<int>& arr){
    int maxEle = -1e8;
    for(int i= 0;i< arr.size(); i++){
        if(arr[i] > maxEle){
            maxEle = arr[i];
        }
    }
    return maxEle;
}

// Find the minmum value
int minmumElement(vector<int>& arr){
    int minEle = 1e8;
    for(int i= 0;i< arr.size(); i++){
        if(arr[i] < minEle){
            minEle = arr[i];
        }
    }
    return minEle;
}

// Find the entered element
int findElement(vector<int>& arr,int data){
    int idx = -1;
    for(int i= 0;i< arr.size(); i++){
        if(arr[i] == data){
            idx = i;
            break;
        }
    }
    return idx;
}

// Find the first index of entered element
int firstIndex(vector<int>& arr,int data){
    int idx = -1;
    for(int i= 0;i< arr.size(); i++){
        if(arr[i] == data){
            idx = i;
            break;
        }
    }
    return idx;
}

// Find the last index of entered element
int lastIndex(vector<int>& arr,int data){
    int idx = -1;
    for(int i= arr.size();i >= 0; i--){
        if(arr[i] == data){
            idx = i;
            break;
        }
    }
    return idx;
}

// Sum of two arrays
void sumOfTwoArrays(vector<int>& arr1, vector<int>& arr2){
    int p = arr1.size();
    int q = arr2.size();
    int r = max(p,q) + 1;

    vector<int> ans(r,0);

    int carry = 0 ;
    int i = p -1;
    int j = q -1;
    int k = r -1;


    while(k >= 0){
        int sum = carry;
        if(i >= 0){
            sum += arr1[i--];
        }    
        if(j >= 0){
            sum += arr2[j--];
        }    

        ans[k--] = sum % 10;
        carry = sum / 10;
    }

    for(int l = 0; l < ans.size() ; l++){
        if( l == 0 && ans[l] == 0){
            continue;
        } 
        cout << ans[l] << endl;
    }
}


// Subtraction of two arrays
void subOfTwoArrays(vector<int>& arr1, vector<int>& arr2){
    int p = arr1.size();
    int q = arr2.size();
    int r = p + 1;

    vector<int> ans(r,0);

    int borrow = 0 ;
    int i = p -1;
    int j = q -1;
    int k = r -1;


    while(k >= 0){
        int num = borrow;
        if(i >= 0){
            num += arr1[i--];
        }    
        if(j >= 0){
            num -= arr2[j--];
        }    
        
        if(num < 0){
            num += 10;
            borrow = -1;
        } else {
            borrow = 0;
        }

        ans[k--] = num;
    }
    
    bool flag = false;
    for(int l = 0; l < ans.size() ; l++){
        if( !flag && ans[l] == 0){
            continue;
        } 
        cout << ans[l] << endl;
        flag = true;
    }
}

// Reverse an array
void reverse(vector<int>& arr, int k){
    int i = 0;
    int j = arr.size() - 1;

    while(i < j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;

        i++;
        j--;
    }
}

// Rotate an array 
void rotate(vector<int>& arr, int k){
    int n = 0;
    // reverse(arr,0,n -1);
}

// Inverse of an array 
void inverse(vector<int>& arr){
        int n = arr.size();
        vector<int> ans(n);

        for(int i = 0; i < ans.size(); i++){
            ans[arr[i]] = i ;
        }

        for(int ele : ans){
            cout << ele << endl;
        }
    }

// Print all subarrays
void printSubArrays(vector<int>& arr){
    for(int i = 0 ; i < arr.size() ; i++){
        for(int j = i ;j < arr.size() ; j++){
            cout << arr[j] << " " ;
        }
        cout << endl;
    }
}

// Binary Search 
int binarySearch(vector<int> arr, int key ){
    int start = 0;
    int end = arr.size() - 1;
    while(start <= end){
        int mid = (start + end) / 2;
        if(key == arr[mid]){
            return mid;
        } 
        else if(key < arr[mid]){
            end = mid - 1;
        }
         else {
            start = mid + 1;
        }
    }
    return -1;
}

// Application of Binary Search Algorithm
// If numbers are arranged in an array in sorted manner in a pair and one digit goes missing 
// You need to find that element/digit 
// Write a function which returns the index of the element 
// Example :- 1 1 2 2 3 3 4 5 5 

int findMissingNumberInPair(vector<int>& arr){
    int size = arr.size() ; 
    if(size % 2 == 0)   return -1;

    int i = 0 ;
    int j = size - 1;
    int mid ;
    while ( i <= j ){
        mid = (i + j) / 2;
        if(i == j) return mid;

        if(mid % 2 == 0){
            if(arr[mid] == arr[mid + 1])    
                i = mid + 2;
            else 
                j = mid;
        } else {
            if(arr[mid] == arr[mid - 1])    
                i = mid + 1;
            else 
                j = mid;
        }
    } 
    return mid; 
}
// ----------------------------------------------------------------

//  Questions .
void input(vector<vector<int>>& arr){
    for(int i= 0; i < arr.size(); i++){
        for(int j=0; j < arr[0].size();j++){
            cin >> arr[i][j];
        }
    }
}
void output(vector<vector<int>>& arr){
    for(int i= 0; i < arr.size(); i++){
        for(int j=0; j < arr[0].size();j++){
            cout << arr[i][j] << " ";
        }
        cout << endl;
    }
}

// Wave print
void waveprint(vector<vector<int>>& arr, int row, int col){
    for(int j = 0; j < col; j++){
        if(j % 2 == 0){
            for(int i = 0 ; i < row ; i++){
                cout << arr[i][j] << " ";
            }
        } else  {
            for(int i = row - 1 ; i >= 0 ; i--){
                cout << arr[i][j] << " ";
            }
        }
    }
}

//  Wave Print from left
void leftwaveprint(vector<vector<int>>& arr, int row, int col){
    for(int i = 0; i < row; i++){
        if(i % 2 == 0){
            for(int j = 0 ; j < col ; j++){
                cout << arr[i][j] << " ";
            }
        } else
        {
            for(int j = col - 1 ; j >= 0 ; j--){
                cout << arr[i][j] << " ";
            }
        }
        
    }
}





// Exit point of array 
void exitPoint(vector<vector<int>>& arr, int row, int col){
    int i = 0, j = 0 , d = 0;

    while (true)        
    {
        d = (d + arr[i][j]) % 4;
        if(d == 0){ // East
            j++;
            if(j == col){
                cout << i << endl << j -1 << endl;
                break;
            }
        } else if(d == 1){ // South
            i++;
            if(j == row){
                cout << i - 1<< endl << j << endl;
                break;
            }
        } else if(d == 2){ // West
            j--;
            if(j == -1){
                cout << i << endl << j + 1 << endl;
                break;
            }
        } else if(d == 3){ // South
            i++;
            if(i == -1){
                cout << i + 1 << endl << j << endl;
                break;
            }
        }

    }
    
}

int main(){
    int n;
    cin >> n;

    vector<int> arr(n);
    for(int i=0; i < arr.size(); i++){
        cin >> arr[i];
    }
    // int key;
    // cin >> key;
    // test2();
    cout << arr[findMissingNumberInPair(arr)];
    return 0;
}